File: com/example/fsrs/Parameter.java
package com.example.fsrs;

public class Parameter {
        private final int maximumInterval;
        private final double requestRetention;
        private final double[] w;
        private final double decay;
        private final double factor;
        private final boolean enableShortTerm;
        private static native long New(
            int maximum_interval,
            double request_retention,
            double[] w,
            double decay,
            double factor,
            boolean enable_short_term);

        public Parameter(int maximumInterval, double requestRetention, double[] w, double decay, double factor,
                boolean enableShortTerm) {
            this.maximumInterval = maximumInterval;
            this.requestRetention = requestRetention;
            this.w = w;
            this.decay = decay;
            this.factor = factor;
            this.enableShortTerm = enableShortTerm;
        }

        public long toNative() {
            return New(maximumInterval, requestRetention, w, decay, factor, enableShortTerm);
        }
    
}

File: src/parameter.rs
use jni::objects::{JClass, JDoubleArray};

use jni::sys::{jboolean, jdouble, jint, jlong};
use jni::JNIEnv;

use crate::{to_raw, Parameter};

#[no_mangle]
pub unsafe extern "system" fn Java_com_example_fsrs_Parameter_New(
    env: JNIEnv,
    _class: JClass,
    maximum_interval: jint,
    request_retention: jdouble,
    w: JDoubleArray,
    decay: jdouble,
    factor: jdouble,
    enable_short_term: jboolean,
) -> jlong {
    let mut buf = vec![0f64; 19];
    env.get_double_array_region(w, 0, &mut buf)
        .expect("too much information");
    to_raw(Parameter {
        inner: fsrs::Parameters {
            maximum_interval,
            request_retention,
            w: buf.try_into().unwrap(),
            decay,
            factor,
            enable_short_term: enable_short_term == (true as u8), // here is u8, 0 or 1
        },
    })
}